### PR TITLE
Optimize tenant, flavor and template selection

### PR DESCRIPTION
**Pull request by nowakowski**
Completely eliminates the n+1 query problem in optimizer.select_tmpl_and_flavor_and_tenant, thus fixing a nasty performance bug in PL-Grid Atmosphere deployment.

### CODE REVIEW

**Comment by nowakowski** [comment]
@mkasztelnik Can you help out with this? It's quite urgent (btw., we probably need an "Urgent" label -- perhaps you could define one).

**Comment by mkasztelnik** [comment]
Can you describe how can I help with it? What is missing? Only review or something more?

**Comment by nowakowski** [comment]
@mkasztelnik I just need a code review.

**Comment by mkasztelnik** [comment]
First of all can you squash all commits into one with meaningful name? Commits like "Interim commit." or "Woof." say nothing about what was changed and why.

File: app/providers/atmosphere/optimization_strategy/default.rb [hunk 1]
```diff
@@ -105,7 +120,7 @@ def select
           best_tenant = nil
           best_flavor = nil
           instantiation_cost = Float::INFINITY
-          tmpls.each do |tmpl|
+          @tmpls.each do |tmpl|
```
Comment (mkasztelnik) [inline app/providers/atmosphere/optimization_strategy/default.rb]: 1 access method should be used in the whole code. So only `@tmpl` and remove `tmpl` getter or use `tmpl` getter. Using both is misleading.

File: app/providers/atmosphere/optimization_strategy/default.rb [hunk 2]
```diff
@@ -187,7 +201,11 @@ def get_optimal_flavor_for_tenant(tmpl, t)
           best_mem = Float::INFINITY
           opt_fl = nil
           mfs.each do |f|
-            present_cost = tmpl.get_hourly_cost_for(f)
+            fofs = f.flavor_os_families
+            osf = fofs.detect do |fof|
```
Comment (mkasztelnik) [inline app/providers/atmosphere/optimization_strategy/default.rb]: Please use more meaningful names

File: spec/services/atmosphere/optimizer_spec.rb [hunk 3]
```diff
@@ -115,8 +115,8 @@
                                   cpu: biggest_os_flavor.cpu,
                                   hdd: biggest_os_flavor.hdd,
                                   tenant: amazon)
-          biggest_os_flavor.set_hourly_cost_for(Atmosphere::OSFamily.first, 100)
-          optimal_flavor.set_hourly_cost_for(Atmosphere::OSFamily.first, 100)
+          biggest_os_flavor.set_hourly_cost_for(Atmosphere::OSFamily.first, 1)
+          optimal_flavor.set_hourly_cost_for(Atmosphere::OSFamily.first, 1)
```
Comment (mkasztelnik) [inline spec/services/atmosphere/optimizer_spec.rb]: Why this change is introduced?

File: app/providers/atmosphere/optimization_strategy/default.rb (hunk 1 above)
Comment (nowakowski) [inline app/providers/atmosphere/optimization_strategy/default.rb]: Uh, I don't believe I understand. `@tmpls` is an instance variable which I'm iterating over. `tmpl` is a local variable which is only needed in the context of this specific loop - hence no `@`. What good would replacing `tmpl` with `@tmpl` do? Seems like a pointless waste of RAM.

File: app/providers/atmosphere/optimization_strategy/default.rb (hunk 2 above)
Comment (nowakowski) [inline app/providers/atmosphere/optimization_strategy/default.rb]: OK.

File: spec/services/atmosphere/optimizer_spec.rb [hunk 4]
```diff
@@ -145,6 +145,25 @@
...
+        vmts << create(
+          :virtual_machine_template,
+          appliance_type: atype,
+          tenants: [t]
+        )
+      end
+      t1 = Time.now.to_f
+      subject.select_tmpl_and_flavor_and_tenant(vmts)
+      t2 = Time.now.to_f
+      expect(t2 - t1).to be < 5
```
Comment (mkasztelnik) [inline spec/services/atmosphere/optimizer_spec.rb]: Please add error message to know what is wrong when this check fails (https://www.relishapp.com/rspec/rspec-expectations/docs/customized-message).

File: app/providers/atmosphere/optimization_strategy/default.rb (hunk 1 above)
Comment (mkasztelnik) [inline app/providers/atmosphere/optimization_strategy/default.rb]: Take a look here https://github.com/dice-cyfronet/atmosphere/blob/optimize_tenant_selection_2/app/providers/atmosphere/optimization_strategy/default.rb#L171. That is why it is misleading.

File: spec/services/atmosphere/optimizer_spec.rb (hunk 3 above)
Comment (nowakowski) [inline spec/services/atmosphere/optimizer_spec.rb]: The old selection logic was bogus. It preferred a flavor with price 100 on amazon over a flavor with price 80 on openstack even when both flavors satisfied AT requirements and the AT had VMTs on both tenants.

This was never caught because (1) the spec was also bogus (hence the change); (2) the situation never came up on PL-Grid.

Btw., I did not write the old selection logic. ;)

Comment (mkasztelnik) [inline spec/services/atmosphere/optimizer_spec.rb]: Can you point me to the place where amazon flavors are created? I'm trying to understand this spec, but it is quite complicated. I come into conclusion that to complicated factories are baaaaaaaaaaad, so many unclear dependencies and assumptions :cry:

File: app/providers/atmosphere/optimization_strategy/default.rb (hunk 1 above)
Comment (nowakowski) [inline app/providers/atmosphere/optimization_strategy/default.rb]: You seem to be confusing `@tmpls` with `@tmpl`. Please re-read your original comment and make sure it doesn't contain any typos in variable names. :)

If you want me to remove `attr_reader :tmpls` then I'll do so. I don't think it's needed anymore.

Comment (mkasztelnik) [inline app/providers/atmosphere/optimization_strategy/default.rb]: You are right `@tmpl` should be `@tmpls` in my original comment - sorry for it - but the comment is still valid.

Comment (mkasztelnik) [inline app/providers/atmosphere/optimization_strategy/default.rb]: E.g. here https://github.com/dice-cyfronet/atmosphere/blob/optimize_tenant_selection_2/app/providers/atmosphere/optimization_strategy/default.rb#L245 and in other places `tmpls` getter is used.

Comment (nowakowski) [inline app/providers/atmosphere/optimization_strategy/default.rb]: Okay, I'll use the getter instead of referring directly to `@tmpls`.

File: spec/services/atmosphere/optimizer_spec.rb (hunk 3 above)
Comment (nowakowski) [inline spec/services/atmosphere/optimizer_spec.rb]: Line 31 - `let(:amazon) { create(:amazon_with_flavors, funds: [fund]) }`.

IMHO default `lets` in specs are baaaaaaaaad but fixing this is out of scope of this PR. :)

File: spec/services/atmosphere/optimizer_spec.rb (hunk 4 above)
Comment (nowakowski) [inline spec/services/atmosphere/optimizer_spec.rb]: Aye.

**Comment by nowakowski** [comment]
I believe I'm done with all requested changes.

**Comment by mkasztelnik** [comment]
:+1: from me. Please merge all commits into one before merge.
